File: output/typescript/tests/09_stack.ts
class StackOpsFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class StackOpsFrameContext {
    public event: StackOpsFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: StackOpsFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class StackOpsCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: StackOpsCompartment | null;

    constructor(state: string, parent_compartment: StackOpsCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): StackOpsCompartment {
        const c = new StackOpsCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class StackOps {
    private _state_stack: Array<any>;
    private __compartment: StackOpsCompartment;
    private __next_compartment: StackOpsCompartment | null;
    private _context_stack: Array<any>;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new StackOpsCompartment("Main");
        this.__next_compartment = null;
        const __frame_event = new StackOpsFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: StackOpsFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new StackOpsFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new StackOpsFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new StackOpsFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: StackOpsFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: StackOpsCompartment) {
        this.__next_compartment = next_compartment;
    }

    public push_and_go() {
        const __e = new StackOpsFrameEvent("push_and_go", null);
        const __ctx = new StackOpsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public pop_back() {
        const __e = new StackOpsFrameEvent("pop_back", null);
        const __ctx = new StackOpsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public do_work(): string {
        const __e = new StackOpsFrameEvent("do_work", null);
        const __ctx = new StackOpsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    public get_state(): string {
        const __e = new StackOpsFrameEvent("get_state", null);
        const __ctx = new StackOpsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    private _state_Main(__e: StackOpsFrameEvent) {
        if (__e._message === "do_work") {
            this._context_stack[this._context_stack.length - 1]._return = "Working in Main";
            return;;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Main";
            return;;
        } else if (__e._message === "pop_back") {
            console.log("Cannot pop - nothing on stack in Main");
        } else if (__e._message === "push_and_go") {
            console.log("Pushing Main to stack, going to Sub");
            this._state_stack.push(this.__compartment.copy());
            const __compartment = new StackOpsCompartment("Sub", this.__compartment.copy());
            this.__transition(__compartment);
        }
    }

    private _state_Sub(__e: StackOpsFrameEvent) {
        if (__e._message === "do_work") {
            this._context_stack[this._context_stack.length - 1]._return = "Working in Sub";
            return;;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Sub";
            return;;
        } else if (__e._message === "pop_back") {
            console.log("Popping back to previous state");
            this.__compartment = this._state_stack.pop()!;
            return;
        } else if (__e._message === "push_and_go") {
            console.log("Already in Sub");
        }
    }
}


function main() {
    console.log("=== Test 09: Stack Push/Pop ===");
    const s = new StackOps();

    // Initial state should be Main
    let state = s.get_state();
    if (state !== "Main") {
        throw new Error(`Expected 'Main', got '${state}'`);
    }
    console.log(`Initial state: ${state}`);

    // Do work in Main
    let work = s.do_work();
    if (work !== "Working in Main") {
        throw new Error(`Expected 'Working in Main', got '${work}'`);
    }
    console.log(`do_work(): ${work}`);

    // Push and go to Sub
    s.push_and_go();
    state = s.get_state();
    if (state !== "Sub") {
        throw new Error(`Expected 'Sub', got '${state}'`);
    }
    console.log(`After push_and_go(): ${state}`);

    // Do work in Sub
    work = s.do_work();
    if (work !== "Working in Sub") {
        throw new Error(`Expected 'Working in Sub', got '${work}'`);
    }
    console.log(`do_work(): ${work}`);

    // Pop back to Main
    s.pop_back();
    state = s.get_state();
    if (state !== "Main") {
        throw new Error(`Expected 'Main' after pop, got '${state}'`);
    }
    console.log(`After pop_back(): ${state}`);

    console.log("PASS: Stack push/pop works correctly");
}

main();

File: output/rust/tests/09_stack.rs
use std::collections::HashMap;

struct StackOpsFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for StackOpsFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl StackOpsFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

struct StackOpsFrameContext {
    event: StackOpsFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl StackOpsFrameContext {
    fn new(event: StackOpsFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct StackOpsCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<StackOpsFrameEvent>,
}

impl StackOpsCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum StackOpsStateContext {
    Main,
    Sub,
    Empty,
}

impl Default for StackOpsStateContext {
    fn default() -> Self {
        StackOpsStateContext::Main
    }
}

pub struct StackOps {
    _state_stack: Vec<(String, StackOpsStateContext)>,
    __compartment: StackOpsCompartment,
    __next_compartment: Option<StackOpsCompartment>,
    _context_stack: Vec<StackOpsFrameContext>,
}

impl StackOps {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            __compartment: StackOpsCompartment::new("Main"),
            __next_compartment: None,
        };
let __frame_event = StackOpsFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: StackOpsFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = StackOpsFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = StackOpsFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = StackOpsFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &StackOpsFrameEvent) {
match self.__compartment.state.as_str() {
    "Main" => self._state_Main(__e),
    "Sub" => self._state_Sub(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: StackOpsCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Main" => StackOpsStateContext::Main,
    "Sub" => StackOpsStateContext::Sub,
    _ => StackOpsStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = StackOpsFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    StackOpsStateContext::Main => {}
    StackOpsStateContext::Sub => {}
    StackOpsStateContext::Empty => {}
}
    }

    pub fn push_and_go(&mut self) {
let mut __e = StackOpsFrameEvent::new("push_and_go");
let __ctx = StackOpsFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Main" => { self._s_Main_push_and_go(&__e); }
            "Sub" => { self._s_Sub_push_and_go(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StackOpsFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StackOpsFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StackOpsFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn pop_back(&mut self) {
let mut __e = StackOpsFrameEvent::new("pop_back");
let __ctx = StackOpsFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Main" => { self._s_Main_pop_back(&__e); }
            "Sub" => { self._s_Sub_pop_back(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StackOpsFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StackOpsFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StackOpsFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn do_work(&mut self) -> String {
let mut __e = StackOpsFrameEvent::new("do_work");
let __ctx = StackOpsFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Main" => { self._s_Main_do_work(&__e); }
            "Sub" => { self._s_Sub_do_work(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StackOpsFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StackOpsFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StackOpsFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<String>().unwrap()
} else {
    Default::default()
}
    }

    pub fn get_state(&mut self) -> String {
let mut __e = StackOpsFrameEvent::new("get_state");
let __ctx = StackOpsFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Main" => { self._s_Main_get_state(&__e); }
            "Sub" => { self._s_Sub_get_state(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StackOpsFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StackOpsFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StackOpsFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<String>().unwrap()
} else {
    Default::default()
}
    }

    fn _state_Sub(&mut self, __e: &StackOpsFrameEvent) {
match __e.message.as_str() {
    "do_work" => { self._s_Sub_do_work(__e); }
    "get_state" => { self._s_Sub_get_state(__e); }
    "pop_back" => { self._s_Sub_pop_back(__e); }
    "push_and_go" => { self._s_Sub_push_and_go(__e); }
    _ => {}
}
    }

    fn _state_Main(&mut self, __e: &StackOpsFrameEvent) {
match __e.message.as_str() {
    "do_work" => { self._s_Main_do_work(__e); }
    "get_state" => { self._s_Main_get_state(__e); }
    "pop_back" => { self._s_Main_pop_back(__e); }
    "push_and_go" => { self._s_Main_push_and_go(__e); }
    _ => {}
}
    }

    fn _s_Sub_pop_back(&mut self, __e: &StackOpsFrameEvent) {
println!("Popping back to previous state");
self._state_stack_pop();
return;
    }

    fn _s_Sub_do_work(&mut self, __e: &StackOpsFrameEvent) {
"Working in Sub".to_string();
    }

    fn _s_Sub_push_and_go(&mut self, __e: &StackOpsFrameEvent) {
println!("Already in Sub");
    }

    fn _s_Sub_get_state(&mut self, __e: &StackOpsFrameEvent) {
"Sub".to_string();
    }

    fn _s_Main_get_state(&mut self, __e: &StackOpsFrameEvent) {
"Main".to_string();
    }

    fn _s_Main_pop_back(&mut self, __e: &StackOpsFrameEvent) {
println!("Cannot pop - nothing on stack in Main");
    }

    fn _s_Main_push_and_go(&mut self, __e: &StackOpsFrameEvent) {
println!("Pushing Main to stack, going to Sub");
self._state_stack_push();
self.__transition(StackOpsCompartment::new("Sub"));
    }

    fn _s_Main_do_work(&mut self, __e: &StackOpsFrameEvent) {
"Working in Main".to_string();
    }
}


fn main() {
    println!("=== Test 09: Stack Push/Pop ===");
    let mut s = StackOps::new();

    // Initial state should be Main
    let state = s.get_state();
    assert_eq!(state, "Main", "Expected 'Main', got '{}'", state);
    println!("Initial state: {}", state);

    // Do work in Main
    let work = s.do_work();
    assert_eq!(work, "Working in Main", "Expected 'Working in Main', got '{}'", work);
    println!("do_work(): {}", work);

    // Push and go to Sub
    s.push_and_go();
    let state = s.get_state();
    assert_eq!(state, "Sub", "Expected 'Sub', got '{}'", state);
    println!("After push_and_go(): {}", state);

    // Do work in Sub
    let work = s.do_work();
    assert_eq!(work, "Working in Sub", "Expected 'Working in Sub', got '{}'", work);
    println!("do_work(): {}", work);

    // Pop back to Main
    s.pop_back();
    let state = s.get_state();
    assert_eq!(state, "Main", "Expected 'Main' after pop, got '{}'", state);
    println!("After pop_back(): {}", state);

    println!("PASS: Stack push/pop works correctly");
}
